### PR TITLE
Bluetooth: Mesh: add metadata to missing models

### PR DESCRIPTION
**Pull request by KyraLengfeld**
This commit adds metadata to the Sensor Server, and Time Server models in form of an optional parameter.

Note: Metadata is only supported if the Large Composition Data Server model is supported.

### CODE REVIEW

File: samples/bluetooth/mesh/light_ctrl/src/model_handler.c [hunk 1]
```diff
@@ -242,11 +260,21 @@ static struct bt_mesh_elem elements[] = {
 	BT_MESH_ELEM(1,
 		     BT_MESH_MODEL_LIST(
 			     BT_MESH_MODEL_CFG_SRV,
-			     BT_MESH_MODEL_HEALTH_SRV(&health_srv, &health_pub),
+#if defined(CONFIG_BT_MESH_LARGE_COMP_DATA_SRV)
+			     BT_MESH_MODEL_HEALTH_SRV(
+					 &health_srv, &health_pub, health_srv_metadata),
+			     BT_MESH_MODEL_LIGHTNESS_SRV(
+					 &my_ctx.lightness_srv, lightness_srv_metadata),
+			     BT_MESH_MODEL_SENSOR_SRV(&sensor_srv, NULL),
```
Comment (alxelax) [inline samples/bluetooth/mesh/light_ctrl/src/model_handler.c:268]: If there is no metadata for the model, it isn't had to be passed in macro. So sensor server can be moved out of ifdef.

File: samples/bluetooth/mesh/light_ctrl/src/model_handler.c [hunk 2]
```diff
@@ -242,11 +260,21 @@ static struct bt_mesh_elem elements[] = {
 	BT_MESH_ELEM(1,
 		     BT_MESH_MODEL_LIST(
 			     BT_MESH_MODEL_CFG_SRV,
-			     BT_MESH_MODEL_HEALTH_SRV(&health_srv, &health_pub),
+#if defined(CONFIG_BT_MESH_LARGE_COMP_DATA_SRV)
```
Comment (alxelax) [inline samples/bluetooth/mesh/light_ctrl/src/model_handler.c:263]: I do not think condition compilation is required for any models except `BT_MESH_MODEL_LARGE_COMP_DATA_SRV`. If there is no Large composition data server, then metadata will go nowhere. There shouldn't be any restrictions to do so.

File: samples/bluetooth/mesh/light_ctrl/src/model_handler.c [hunk 3]
```diff
@@ -101,6 +102,23 @@ static const struct bt_mesh_health_srv_cb health_srv_cb = {
 	.attn_off = attention_off,
 };
 
+#ifdef CONFIG_BT_MESH_LARGE_COMP_DATA_SRV
```
Comment (alxelax) [inline samples/bluetooth/mesh/light_ctrl/src/model_handler.c:105]: `ifdef` is not required

**Review by alxelax — CHANGES_REQUESTED** [review]
I guess release notes should be updated too.

File: samples/bluetooth/mesh/light_ctrl/src/model_handler.c (hunk 2 above)
Comment (KyraLengfeld) [reply samples/bluetooth/mesh/light_ctrl/src/model_handler.c:263]: Will give it a try :)

File: samples/bluetooth/mesh/light_ctrl/src/model_handler.c (hunk 3 above)
Comment (KyraLengfeld) [reply samples/bluetooth/mesh/light_ctrl/src/model_handler.c:105]: Will remove if [that](https://github.com/nrfconnect/sdk-nrf/pull/16580/files#r1688171790) works. (Otherwise it will give compilation warnings when not using large comp data, I guess)

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 4]
```diff
@@ -107,7 +107,10 @@ Bluetooth® LE
 Bluetooth Mesh
 --------------
 
-|no_changes_yet_note|
+* Updated:
+
+  *  Metadata has been added to the Light Lightness Server, Light HSL Server, Light CTL Temperature Server, the Sensor Server, and Time Server models in form of an optional parameter.
+     Usage is demonstrated in the light_ctrl sample.
```
Comment (alxelax) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:113]: I think it is worth to add reference to sample documentation here.

File: include/bluetooth/mesh/sensor_srv.h [hunk 5]
```diff
@@ -44,20 +44,36 @@ struct bt_mesh_sensor_srv;
  *  @brief Sensor Server model composition data entry.
  *
  *  @param[in] _srv Pointer to a @ref bt_mesh_sensor_srv instance.
+ *  @param ...      Optional Light Lightness Server metadata if application is
```
Comment (alxelax) [inline include/bluetooth/mesh/sensor_srv.h:47]:
```suggestion
 *  @param ...      Optional Light Lightness server metadata if application is
```

File: include/bluetooth/mesh/time_srv.h [hunk 6]
```diff
@@ -46,19 +46,46 @@ struct tm;
  * @brief Time Server model composition data entry.
  *
  * @param[in] _srv Pointer to a @ref bt_mesh_time_srv instance.
+ * @param ...	   Optional Time Server metadata if application is
+ *		   compiled with Large Composition Data Server support,
```
Comment (alxelax) [inline include/bluetooth/mesh/time_srv.h:50]:
```suggestion
 *		   compiled with Large Composition Data server support,
```

**Review by alxelax — COMMENTED** [review]
It looks better but still requires some changes.

**Comment by alxelax** [comment]
Also, it requires some tech writer review. Is @mia-ko available?

File: include/bluetooth/mesh/time_srv.h (hunk 6 above)
Comment (KyraLengfeld) [reply include/bluetooth/mesh/time_srv.h:50]: Will align this in the other param descriptions as well (light servers) then.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 4 above)
Comment (omkar3141) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:113]: Needs `:ref:` link.

File: include/bluetooth/mesh/sensor_srv.h (hunk 5 above)
Comment (omkar3141) [reply include/bluetooth/mesh/sensor_srv.h:47]: Why? Everywhere 'Server' is capitalized. [See](https://github.com/nrfconnect/sdk-zephyr/blob/38452400d3bed75faf30b5a0e4a5a34d9a44128b/include/zephyr/bluetooth/mesh/health_srv.h#L169)

File: include/bluetooth/mesh/time_srv.h (hunk 6 above)
Comment (omkar3141) [reply include/bluetooth/mesh/time_srv.h:50]: Why? 'Server' is capitalized. [See](https://github.com/nrfconnect/sdk-zephyr/blob/38452400d3bed75faf30b5a0e4a5a34d9a44128b/include/zephyr/bluetooth/mesh/health_srv.h#L169)

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 7]
```diff
@@ -107,7 +107,10 @@ Bluetooth® LE
 Bluetooth Mesh
 --------------
 
-|no_changes_yet_note|
+* Updated:
+
+  *  Metadata has been added to the Light Lightness Server, Light HSL Server, Light CTL Temperature Server, the Sensor Server, and Time Server models in form of an optional parameter.
```
[the remark this answers is not in the document]
Comment (omkar3141) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:112]: AFAIK it is capitalized almost everywhere where "model" follows the term.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 4 above)
Comment (KyraLengfeld) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:113]: I'll add a sentence under ´Bluetooth samples´ section.

File: samples/bluetooth/mesh/light_ctrl/README.rst [hunk 8]
```diff
@@ -317,6 +323,7 @@ This sample uses the following |NCS| libraries:
 * :ref:`bt_mesh_light_ctrl_srv_readme`
 * :ref:`bt_mesh_scene_srv_readme`
 * :ref:`bt_mesh_sensor_srv_readme`
+* :ref:`bt_mesh_large_comp_data_srv_readme`
```
Comment (alxelax) [inline samples/bluetooth/mesh/light_ctrl/README.rst:326]:
```suggestion
* :ref:`zephyr:bluetooth_mesh_lcd_srv`
```
The Large Composition Data server is part of Mesh core functionality in upstream.

File: samples/bluetooth/mesh/light_ctrl/README.rst [hunk 9]
```diff
@@ -96,6 +98,8 @@ The following table shows the mesh light fixture composition data for this sampl
    +-------------------------------+----------------------------+
    | Sensor Setup Server           |                            |
    +-------------------------------+----------------------------+
+   | Large Comp Data Server        |                            |
```
Comment (omkar3141) [inline samples/bluetooth/mesh/light_ctrl/README.rst:101]: I thought you were just using `light_ctrl` for your own testing purpose. I'm really uncertain about inserting LCD Server in this sample. This sample is our "the BLC NLC and ENM NLC sample". None of those specs mandate usage of LCD server. So, adding it here will simply bloat the sample and make this sample a tad bit more complicated.

I would highly recommend to not add metadata into this sample.

Instead of this, could we rather add these metadata in LCD test, mesh_shell, and in PTS tester, like done for health server? These things could be done as a seperate task.

File: samples/bluetooth/mesh/light_ctrl/prj.conf [hunk 10]
```diff
@@ -52,10 +52,11 @@ CONFIG_BT_MESH_DK_PROV=y
 CONFIG_BT_MESH_NLC_PERF_CONF=y
 CONFIG_BT_MESH_NLC_PERF_LIGHTNESS_CTRL=y
 CONFIG_BT_MESH_MODEL_EXTENSIONS=y
-CONFIG_BT_MESH_MODEL_EXTENSION_LIST_SIZE=18
+CONFIG_BT_MESH_MODEL_EXTENSION_LIST_SIZE=19
 # Enabling BT_MESH_NLC_PERF_CONF enables support for 3 application keys by
 # default. Therefore, allow up to 3 application key bindings per model instance.
 CONFIG_BT_MESH_MODEL_KEY_COUNT=3
+CONFIG_BT_MESH_LARGE_COMP_DATA_SRV=y
```
Comment (omkar3141) [inline samples/bluetooth/mesh/light_ctrl/prj.conf:59]: I would recommend to revert these changes.

**Review by omkar3141 — CHANGES_REQUESTED** [review]

**Comment by KyraLengfeld** [comment]
BTW I noticed there is no large composition data server readme, as there is with for example the light lightness control server (`doc/nrf/libraries/bluetooth_services/mesh/light_ctrl_srv.rst`). Is this intentional?

File: include/bluetooth/mesh/time_srv.h (hunk 6 above)
Comment (alxelax) [reply include/bluetooth/mesh/time_srv.h:50]: There are about half of them capitalized and half are not. Then up to implementer what to do with this. I do not think this is really important to argue about this.

**Comment by alxelax** [comment]
> BTW I noticed there is no large composition data server readme, as there is with for example the light lightness control server (`doc/nrf/libraries/bluetooth_services/mesh/light_ctrl_srv.rst`). Is this intentional?

Please look at my comment: https://github.com/nrfconnect/sdk-nrf/pull/16580#discussion_r1689396480

File: samples/bluetooth/mesh/light_ctrl/README.rst (hunk 9 above)
Comment (KyraLengfeld) [reply samples/bluetooth/mesh/light_ctrl/README.rst:101]: Changed light_ctrl sample, because that was part of the DoD in the task :) but we can discuss this when I am in the office after lunch more? Will update a cleaned-up version of this PR after.

File: samples/bluetooth/mesh/light_ctrl/prj.conf (hunk 10 above)
Comment (omkar3141) [reply samples/bluetooth/mesh/light_ctrl/prj.conf:59]: It is in a same way like Health Server. Where we don't add metadata for health server in every sample. That is fully tested via tests. We should perhaps do the same and keep this sample compact and not add non-required features here. This is our showcase sample for BLCNLCP + ENMNLCP profiles and they do not mandate metadata.

File: include/bluetooth/mesh/time_srv.h (hunk 6 above)
Comment (KyraLengfeld) [reply include/bluetooth/mesh/time_srv.h:50]: I'll align capitalization where I find it (I.e. leave them capitalized and amend the times it is not). Agree it is not that important, but nice to have it same everywhere.

File: samples/bluetooth/mesh/light_ctrl/README.rst (hunk 8 above)
Comment (KyraLengfeld) [reply samples/bluetooth/mesh/light_ctrl/README.rst:326]: that explains it! thanks :)

File: samples/bluetooth/mesh/light_ctrl/README.rst (hunk 9 above)
Comment (alxelax) [reply samples/bluetooth/mesh/light_ctrl/README.rst:101]: No idea. @omkar3141 initially created the task. Seems he disagrees at the moment to what he thought about this in time of task creation. :)
I'm Ok to not add it in sample either. I'm not sure if we have to add metadata artificially to anywhere at all. If we need to check whether it works or doesn't just use unit testing.

Comment (KyraLengfeld) [reply samples/bluetooth/mesh/light_ctrl/README.rst:101]: and @omkar3141 if you look at the [task description](https://nordicsemi.atlassian.net/browse/NCSDK-22500), one goal was to add it to a sample, not just mesh_shell.

File: samples/bluetooth/mesh/light_ctrl/prj.conf (hunk 10 above)
Comment (alxelax) [reply samples/bluetooth/mesh/light_ctrl/prj.conf:59]: Then we do not need to add metadata in sample

**Comment by KyraLengfeld** [comment]
changelog will be added in another task once @mia-ko is back from vacation.

**Review by alxelax — APPROVED** [review]

**Review by omkar3141 — APPROVED** [review]
